File: 53/53.c
#include <stdio.h>
#include <string.h>

int maxSubArray(int* nums, int numsSize) {
    int current_max_sum = 0, max_sum = nums[0];
    size_t idx;
    /* for (idx = 0; idx < (size_t)numsSize; ++idx) { */
    /*     current_max_sum = current_max_sum & 0x80000000 ? nums[idx] : current_max_sum + nums[idx]; */
    /*     printf("%d  %d\n", current_max_sum, max_sum); */
    /*     max_sum = current_max_sum > max_sum ? current_max_sum : max_sum; */
    /* } */
    for (idx = 0; idx < (size_t)numsSize; ++idx) {
        current_max_sum += nums[idx];
        /* max_sum = current_max_sum > max_sum ? current_max_sum : max_sum; */
        if (current_max_sum > max_sum) {
            max_sum = current_max_sum;
        }
        /* current_max_sum = current_max_sum & 0x80000000 ? 0 : current_max_sum; */
        if (current_max_sum & 0x80000000) { // if faster then = ? :, maybe due to branch prediction
            current_max_sum = 0;
        }
    }
    return max_sum;
}

int main() {
    int nums[] = {-3,-2,-2,-3};
    printf("%d", maxSubArray(nums, 4));
    return 0;
}

File: 53/53.rs
struct Solution {}

impl Solution {
    pub fn max_sub_array(nums: Vec<i32>) -> i32 {
        let mut sum_max_so_far: i32 = 0;
        let mut max_sum: i32 = nums[0];
        for n in nums.iter() {
            if (sum_max_so_far < 0) {
                sum_max_so_far = 0;
            }
            sum_max_so_far += n;
            max_sum = std::cmp::max(max_sum, sum_max_so_far);
        }
        max_sum
    }
    // pub fn max_sub_array(nums: Vec<i32>) -> i32 {
    //     let mut sum_max_so_far: i32 = 0;
    //     let mut max_sum: i32 = nums[0];
    //     for n in nums.iter() {
    //         sum_max_so_far += n;
    //         if (max_sum < sum_max_so_far) {
    //             max_sum = sum_max_so_far;
    //         }
    //         if (sum_max_so_far < 0) {
    //             sum_max_so_far = 0;
    //         }
    //     }
    //     max_sum
    // }
}
